src/build.d

```diff
@@ -441,6 +448,51 @@ auto dmdUnittestExe = dmdExe("-unittest", ["-version=NoMain", "-unittest", "-mai
         .command(dmdUnittestExe.targets);
 });
 
+/**
+BuildRule to run the DMD frontend header generation
+For debugging, use `./build.d cxx-headers DFLAGS="-debug=Debug_DtoH"` (clean before)
+*/
+alias buildFrontendHeaders = makeRule!((builder, rule) {
+    const dmdSources = sources.dmd.all ~ sources.root ~ sources.lexer;
+    const dmdExeFile = dmdDefault.deps[0].target;
+    builder
+        .name("cxx-headers")
+        .description("Build the C++ frontend headers ")
+        .msg("(RUN) CXX-HEADERS")
+        .deps([dmdDefault])
+        .target(env["G"].buildPath("frontend.h"))
+        .command([dmdExeFile] ~ flags["DFLAGS"] ~
+            ["-J" ~ env["RES"], "-c", "-o-", "-HCf="~rule.target] ~ dmdSources);
+});
+
+alias runCxxHeadersTest = makeRule!((builder, rule) {
+    builder
+        .name("cxx-headers-test")
+        .description("Test the C++ frontend headers ")
+        .msg("(TEST) CXX-HEADERS")
+        .deps([buildFrontendHeaders])
+        .commandFunction(() {
+            const cxxHeaderGeneratedPath = buildFrontendHeaders.target;
+            const cxxHeaderReferencePath = srcDir.buildPath("frontend.h");
+            log("Comparing referenceHeader(%s) <-> generatedHeader(%s)",
+                cxxHeaderReferencePath, cxxHeaderGeneratedPath);
+            const generatedHeader = cxxHeaderGeneratedPath.readText;
+            const referenceHeader = cxxHeaderReferencePath.readText;
+            if (generatedHeader != referenceHeader) {
+                string message = "ERROR: Newly generated header file (" ~ cxxHeaderGeneratedPath ~
+                    ") doesn't match with the reference header file (" ~
+                    cxxHeaderReferencePath ~ ")";
+                version(Posix)
+                {
+                    auto diff = tryRun(["diff", "-u", cxxHeaderReferencePath, cxxHeaderGeneratedPath], runDir).output;
+                    abortBuild(message, diff);
+                } else {
+                    abortBuild(message);
```

Comment:
There's already a check doing so:

https://github.com/dlang/dmd/blob/fd295cf5793bc0362947e875965908e93cfd0199/.circleci/run.sh#L153-L164

So we could drop this target in favor of this and change the earlier target to overwrite `frontend.h` directly. I thought this is more explicit and we can convey to the developer better why the CI fails.

Reply:
Phrased that one poorly, i meant using `git` to diff the existing and the generated file (hence the `--no-index`).

```d
auto diff = tryRun(["git", "diff", "--no-index", cxxHeaderReferencePath, cxxHeaderGeneratedPath], runDir).output;
```

Reply:
Oh, fair enough (though at the moment the frontend header generation doesn't seem to work on either Windows nor OSX).
Changed in https://github.com/dlang/dmd/pull/11575/commits/53dc579d69a6adb509a7dc66fb2f53999ea84c09

> So we could drop this target in favor of this and change the earlier target to overwrite frontend.h directly.

What do you think about this btw? It might be easier for people to update as then it's just `./src/build.d cxx-headers` and commit the changes.

Reply:
Not sure about that one. Personally not a fan of generating files outside of the build folder because it can overwrite other changes by accident (especially if the target already exists).

Reply:
On a second thought, `build.d` could update the file automatically when `AUTO_UPDATE=1` is passed on the command line/environment.

Reply:
> On a second thought, build.d could update the file automatically when AUTO_UPDATE=1 is passed on the command line/environment.

Added that option.